src/collection.cpp

```diff
@@ -0,0 +1,126 @@
+#include <morphio/collection.h>
+
+#include "shared_utils.hpp"
+#include <highfive/H5File.hpp>
+
+namespace morphio {
+
+namespace detail {
+template <class Derived>
+class CollectionImpl: public morphio::CollectionImpl
+{
+    // The purpose of this class is to implement the two separate
+    // function in terms of a single templated method `load_impl`.
+  public:
+    morphio::Morphology load(const std::string& morph_name) const override {
+        const auto& derived = static_cast<const Derived&>(*this);
+        return derived.template load_impl<Morphology>(morph_name);
+    }
+
+    morphio::mut::Morphology load_mut(const std::string& morph_name) const override {
+        const auto& derived = static_cast<const Derived&>(*this);
+        return derived.template load_impl<mut::Morphology>(morph_name);
+    }
+};
+}  // namespace detail
+
+class DirectoryCollection: public morphio::detail::CollectionImpl<DirectoryCollection>
+{
+  public:
+    DirectoryCollection(std::string collection_path, std::vector<std::string> extensions)
+        : _dirname(std::move(collection_path))
+        , _extensions(extensions) {}
+
+  protected:
+    friend morphio::detail::CollectionImpl<DirectoryCollection>;
+
+    template <class M>
+    M load_impl(const std::string& morph_name) const {
+        return M(morphology_path(morph_name));
+    }
+
+    std::string morphology_path(const std::string& morph_name) const {
+        for (const auto& ext : _extensions) {
+            auto path = morphio::join_path(_dirname, morph_name + "." + ext);
+            if (morphio::is_regular_file(path)) {
+                return path;
+            }
+        }
+
+        throw std::runtime_error("Morpholgy '" + morph_name + "' not found in: " + _dirname);
+    }
+
+  private:
+    std::string _dirname;
+    std::vector<std::string> _extensions;
+};
+
+class ContainerCollection: public morphio::detail::CollectionImpl<ContainerCollection>
```

Comment:
could we add HDF5 to the name, just so it's clear what the backing store is?

Reply:
done.